<a href="https://colab.research.google.com/github/mudesir/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Mudesir_of_DS16_Unit_1_Sprint_Challenge_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 2

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 is a variable that indicates whether or not a person's request for credit has been approved or denied. This is a good candidate for a y variable since we might want to use the other features to predict this one. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [0]:
# load the data
import pandas as pd
import numpy as np
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data')
print(df.shape)
df.head()

(689, 16)


b  30.83      0  u  g  w  v  1.25  t t.1  01  f g.1  00202  0.1  +
0  a  58.67  4.460  u  g  q  h  3.04  t   t   6  f   g  00043  560  +
1  a  24.50  0.500  u  g  q  h  1.50  t   f   0  f   g  00280  824  +
2  b  27.83  1.540  u  g  w  v  3.75  t   t   5  t   g  00100    3  +
3  b  20.17  5.625  u  g  w  v  1.71  t   f   0  f   s  00120    0  +
4  b  32.08  4.000  u  g  m  v  2.50  t   f   0  t   g  00360    0  +

In [0]:
column_headers = ['A1', 'A2','A3','A4','A5', 'A6','A7','A8','A9', 'A10','A11','A12','A13', 'A14','A15','A16']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data', names=column_headers)
print(df.shape)
df.head()

(690, 16)


A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

In [0]:
df.isnull().sum()

A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

In [0]:
df = df.replace({'+':1, '-': 0, '?': np.NaN})
print(df.shape)
df.head()

(653, 16)


A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15  A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0    1
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560    1
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824    1
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3    1
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0    1

In [0]:
df.dropna(axis=0, how='any', inplace=True)
df.isnull().sum()


A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

In [0]:
df.dtypes

A1      object
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14     object
A15      int64
A16      int64
dtype: object

In [0]:
df['A2'].dropna(inplace=True)
df = df.replace({'00000': 0, })

In [0]:
df['A14'].dropna(inplace=True)
df['A14'].isnull().sum()
#type(df['A2'])
#df['A2']

0

In [0]:
df.describe()

A3          A8         A11            A15         A16
count  653.000000  653.000000  653.000000     653.000000  653.000000
mean     4.829533    2.244296    2.502297    1013.761103    0.453292
std      5.027077    3.371120    4.968497    5253.278504    0.498195
min      0.000000    0.000000    0.000000       0.000000    0.000000
25%      1.040000    0.165000    0.000000       0.000000    0.000000
50%      2.835000    1.000000    0.000000       5.000000    0.000000
75%      7.500000    2.625000    3.000000     400.000000    1.000000
max     28.000000   28.500000   67.000000  100000.000000    1.000000

In [0]:
df.describe(exclude='number')

A1     A2   A4   A5   A6   A7   A9  A10  A12  A13  A14
count   653    653  653  653  653  653  653  653  653  653  653
unique    2    340    3    3   14    9    2    2    2    3  164
top       b  22.67    u    g    c    v    t    f    f    g    0
freq    450      9  499  499  133  381  349  366  351  598  128

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [0]:
# I already changed + with 1 and - with 0 for A16
ctdata = pd.crosstab(df['A16'],df['A8'])
ctdata

A8   0.000   0.040   0.085   0.125   ...  17.500  18.000  20.000  28.500
A16                                  ...                                
0        46      19      19      24  ...       0       0       0       0
1        14      13       5       5  ...       1       1       2       1

[2 rows x 131 columns]

1. Null Hypothesis the two means are equal
2. Alternative the two means are not equal
3. Confidence Interval: 95%


In [0]:
b_acept = df[df['A16']==1]
b_rejec = df[df['A16']==0]
b_acept

A1     A2     A3 A4 A5  A6 A7    A8 A9 A10  A11 A12 A13    A14   A15  A16
0    b  30.83  0.000  u  g   w  v  1.25  t   t    1   f   g  00202     0    1
1    a  58.67  4.460  u  g   q  h  3.04  t   t    6   f   g  00043   560    1
2    a  24.50  0.500  u  g   q  h  1.50  t   f    0   f   g  00280   824    1
3    b  27.83  1.540  u  g   w  v  3.75  t   t    5   t   g  00100     3    1
4    b  20.17  5.625  u  g   w  v  1.71  t   f    0   f   s  00120     0    1
..  ..    ...    ... .. ..  .. ..   ... ..  ..  ...  ..  ..    ...   ...  ...
597  a  21.50  6.000  u  g  aa  v  2.50  t   t    3   f   g  00080   918    1
599  b  20.50  2.415  u  g   c  v  2.00  t   t   11   t   g  00200  3000    1
600  a  29.50  0.460  u  g   k  v  0.54  t   t    4   f   g  00380   500    1
606  b  16.17  0.040  u  g   c  v  0.04  f   f    0   f   g      0     0    1
621  b  22.67  0.165  u  g   c  j  2.25  f   f    0   t   s      0     0    1

[296 rows x 16 columns]

In [0]:
from scipy import stats
stats.ttest_ind(b_rejec['A8'], b_acept['A8'])

Ttest_indResult(statistic=-9.002392498622463, pvalue=2.4079238505859142e-18)

To chi2 test
1. Null hypothesis the two variables are independent
2. Alternative the two variables are not independent
3. Confidence level: 95%

In [0]:
chi2, P_value, dof, expected = stats.chi2_contingency(ctdata)
print('chi2 ', chi2)
print('P-value ', P_value)
print('dof ', dof)
print('expected ', expected)


chi2  211.11376612716873
P-value  8.861958022514069e-06
dof  130
expected  [[32.80245023 17.49464012 13.12098009 15.85451761 10.93415008  3.28024502
  17.49464012  6.56049005  2.73353752  3.82695253  4.37366003  0.5467075
   0.5467075  14.7611026   2.73353752  3.28024502  1.09341501  3.28024502
   0.5467075   6.56049005  0.5467075   0.5467075   2.18683002  2.18683002
   1.09341501 16.40122511  0.5467075   3.28024502  2.18683002  1.09341501
   5.46707504  1.64012251  1.09341501  1.09341501  1.64012251  1.09341501
  13.12098009  0.5467075   2.73353752  2.73353752  1.64012251  0.5467075
   8.20061256  1.09341501  0.5467075   0.5467075   8.20061256  0.5467075
   1.09341501  0.5467075   0.5467075   4.92036753  1.64012251  0.5467075
   1.09341501  1.09341501  0.5467075   9.29402757  0.5467075   1.09341501
   1.09341501  1.64012251  0.5467075   5.46707504  1.09341501  1.64012251
   0.5467075   1.09341501  0.5467075   1.64012251  0.5467075   6.01378254
   1.64012251  0.5467075   4.37366003  0.

## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

* Conclusion t-tes
due to a t-statisc of - 9.0 and p-value 0, we reject the null hypothesis that the two mean are equal. We suggest the alternative the two means are not equal.*
*conclusion for chi2 test
Based on a chi2 statistcs of 211.11 and a p- value of 0.000009 we reject the null hypothesis the two variables are independent we suggest the alternative the two variabls are dependent.

## Part 4 - Bayesian vs Frequentist Statistics

Using a minimum of 2-3 sentences, give an explanation of Bayesian and Frequentist statistics, and then compare and contrast these two approaches to statistical inference.



The basian statistics is a system of experssing uncertainity not depending any other sample , but frequentist are based on sample and inferencing

# Stretch Goals: 
Do these to get a 3. These are not required in order to pass the Sprint Challenge.

## Part 1: 

Make sure that all of your dataframe columns have the appropriate data types. *Hint:* If a column has the datatype of "object" even though it's made up of float or integer values, you can coerce it to act as a numeric column by using the `pd.to_numeric()` function. In order to get a 3 on this section make sure that your data exploration is particularly well commented, easy to follow, and thorough.

## Part 2:

Write functions that can calculate t-tests and chi^2 tests on all of the appropriate column combinations from the dataset. (Remember that certain tests require certain variable types.)

## Part 3: 

Calculate and report confidence intervals on your most important mean estimates (choose at least two). Make some kind of a graphic or visualization to help us see visually how precise these estimates are.

## Part 4:

Give an extra awesome explanation of Bayesian vs Frequentist Statistics. Maybe use code or visualizations, or any other means necessary to show an above average grasp of these high level concepts.

In [0]:
# You can work the stretch goals down here or back up in their regular sections
# just make sure that they are labeled so that we can easily differentiate
# your main work from the stretch goals.